src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs

```diff
@@ -422,6 +424,15 @@ private static NamedTypeSymbol ValidateBuilderType(SyntheticBoundNodeFactory F,
                     !property.IsStatic &&
                     (property.ParameterCount == 0))
                 {
+                    if (property.Type != returnType)
```

Comment:
The equality is checking custom modifiers used within the type by default. However, there is another  portion of them on the property itself, that is not checked by this code. I think we should compare these types in "ignore custom modifiers" mode. Custom modifiers are significant only in call-site signatures and overriding/implementing, they should be ignored otherwise. I leave it up to you to decide if we should ignore dynamic/tuple names mismatch.